spec/libraries/duplicator_spec.rb

```diff
@@ -0,0 +1,561 @@
+        duplicator = Duplicator.new([@sar_1])
+        dup_sar_1 = duplicator.duplicate(@sar_1)
+
+        expect(dup_sar_1).to be_nil
+        expect(SimpleActiveRecord.count).to be(1)
+      end
+
+      it 'is duplicated once' do
+        duplicator = Duplicator.new
+        duplicator.duplicate(@sar_1).save
+        duplicator.duplicate(@sar_1).save
+
+        expect(SimpleActiveRecord.count).to be(2)
+      end
+    end
+  end
+
+  # ComplexActiveRecord objects have associations to themselves
+  context 'when ComplexActiveRecord objects are duplicated' do
+    #self.use_transactional_fixtures = false
+    with_temporary_table(:complex_active_records) do
+      with_temporary_table(:children_parents) do
+        it 'duplicates a ComplexActiveRecord object' do
+          # create object and some children
+          c1 = ComplexActiveRecord.new(data: 11)
+          c1.save
+          c1.children.create({data: 12})
+          c1.children.create({data: 13})
+
+          # duplicate object
+          duplicator = Duplicator.new
+          dup_c1 = duplicator.duplicate(c1)
+          dup_c1.save
+
+          # tests
+          expect(ComplexActiveRecord.count).to be(6)
+          expect(c1).to eq(dup_c1)
+          expect(c1).to_not be(dup_c1)
+          # both children should be duplicated, same values, different objects
+          (0..1).each do |i|
+            expect(c1.children[i]).to eq(dup_c1.children[i])
+            expect(c1.children[i]).to_not be(dup_c1.children[i])
+          end
+        end
+
+        it 'duplicates object referenced in 2 places once' do
+          #
+          # c1--|
+          #     |---> c5
+          # c2--|
+          #
+          c1 = ComplexActiveRecord.new(data: 11)
+          c2 = ComplexActiveRecord.new(data: 12)
+          c1.save
+          c2.save
+          c1.children.create({data: 15})
+          c2.children = c1.children
+
+          duplicator = Duplicator.new
+          dup_c1 = duplicator.duplicate(c1)
+          dup_c1.save
+          dup_c2 = duplicator.duplicate(c2)
+          dup_c2.save
+
+          expect(ComplexActiveRecord.count).to be(6)
+          expect(dup_c1.children[0]).to be(dup_c2.children[0])
+          expect(dup_c1.children[0].data).to be(15)
+        end
+
+        # stuck here because c3 never gets duplicated
+        it 'duplicates multi-layer graphs' do
+          #
+          #  c1 ---> c2 ---- > c3
+          #     ----> c4 ----> c5
+          #     ----> c6
+          #
+          c1 = ComplexActiveRecord.new(data: 11)
+          c1.save
+          c2 = c1.children.create({data: 12})
+          c3 = c2.children.create({data: 13})
+          c4 = c1.children.create({data: 14})
+          c5 = c4.children.create({data: 15})
+          c6 = c1.children.create({data: 16})
+#          c6.children = c4.children
+
+          duplicator = Duplicator.new
+          dup_c1 = duplicator.duplicate(c1)
+          dup_c1.save
+          p dup_c1.children[0].children[0]
```

Comment:
@lowjoel this fails because the child of duplicated c2 has mysteriously become c5. can't figure out why :confused: Thus c3 is never duplicated.

```
c1 --> c2 --> c3
```
is ok.

It fails at
```
c1--> c2 --> c3
   --> c4 --> c5
```
when there are 2 parallel multi-layered paths.

The orig graph is correct (checked in database). But when it goes to the duplicate function, the child of c2 is c5.